### PR TITLE
ENH: Profile memory

### PR DESCRIPTION
**Pull request by larsoner**
This PR:

- Rebased #374 
- fixed a sneaky problem where `memory_profiler.memory_usage` could end up calling a given function more than once (!)
- added docs for `show_memory`
- adds `memory_profiler` to `.circleci`
- modifies the call to use a "memory delta" rather than the total memory (which could accumulate due to running)

Ready for review/merge from my end. @GaelVaroquaux you had the most recent comment on this.

@tamasgal do you want to see if it works for you?

### CODE REVIEW

**Comment by larsoner** [comment]
Example output:

https://587-25860190-gh.circle-artifacts.com/0/html/auto_examples/plot_exp.html#sphx-glr-auto-examples-plot-exp-py

**Comment by larsoner** [comment]
So I realized that `memory_profiler.memory_usage` does the memory usage of the running Python process, which is probably not so good because then the variables being put into memory for the sphinx-build affect the memory shown for examples, which is not so good. Instead we probably want a "memory delta" which is the additional memory caused by running the example. So I've changed it to use that.

The disadvantage is that the memory estimates do not include the startup memory cost of a Python interpreter, so they are systematically low. But for examples where the memory usage is most useful -- ones that use a lot of memory -- they should still be useful. Not sure if there is a better approach here.

But this is what the rendered version now looks like:

https://588-25860190-gh.circle-artifacts.com/0/rtd_html/auto_examples/plot_seaborn.html#sphx-glr-auto-examples-plot-seaborn-py

**Comment by tamasgal** [comment]
Looks good to me!

**Comment by larsoner** [comment]
I have added a base estimate of the memory by using `subprocess.Popen`. It's not so pretty but should work.

File: sphinx_gallery/tests/test_gen_rst.py [hunk 1]
```diff
@@ -296,7 +296,7 @@ def test_rst_example(gallery_conf):
                                 })
 
     example_file = os.path.join(gallery_conf['gallery_dir'], "plot.py")
-    sg.save_rst_example("example_rst", example_file, 0, gallery_conf)
+    sg.save_rst_example("example_rst", example_file, 0, [], gallery_conf)
```
[the remark this answers is not in the document]
Comment (larsoner) [reply sphinx_gallery/tests/test_gen_rst.py]: In the `gallery_conf` `show_memory=False` so it's never used. But I'll change it so it's at least a valid arg

**Comment by larsoner** [comment]
Refactoring complete @Titan-C 

https://595-25860190-gh.circle-artifacts.com/0/rtd_html/auto_examples/plot_exp.html

**Comment by Titan-C** [comment]
Great @larsoner, thank you for working on this. +1 on my side, ready for merge

**Pull request merged**